### PR TITLE
List.find_map : ('a -> 'b option) -> 'a list -> 'b option

### PR DESCRIPTION
**Pull request by gasche**
Edit: the original name proposed was `List.find_some`, but I edited the PR title to follow the current naming consensus.

### CODE REVIEW

File: stdlib/list.ml
Comment (nojb) [inline stdlib/list.ml]: Can you please indent this in a reasonable way? Thanks!

File: stdlib/list.mli
Comment (nojb) [inline stdlib/list.mli:239]: Please, only one line of whitespace between declarations.

File: stdlib/list.mli
Comment (yallop) [inline stdlib/list.mli]: I prefer "each element" to "every element" here, since the traversal stops once `f` "succeeds".

**Review by nojb — APPROVED** [review]
LGTM (modulo trivial remarks). Thanks for your new-found zest for stdlib additions!

**Comment by Armael** [comment]
~~Note that both Containers and Batteries use `List.find_opt` as a name for this function.~~
~~It might be worth sticking to that name if that's what people are already used to.~~
Nvm!

File: stdlib/list.mli
Comment (alainfrisch) [inline stdlib/list.mli]:
```suggestion
(** [find_some f l] applies [f] successively to elements of [l],
```

might remove the wrong impression that the function is indeed applied to all elements.

**Comment by yallop** [comment]
This is a useful function to have, but I'm not very keen on the name, which suggests (to me) that the function is looking for a value of the form `Some v` rather than a value for which `f` returns `Some v`.

**Comment by nojb** [comment]
> Note that both Containers and Batteries use `List.find_opt` as a name for this function.
> It might be worth sticking to that name if that's what people are already used to.

`List.find_opt` would surely have a different signature: `('a -> bool) -> 'a list -> 'a option`, right?

**Comment by nojb** [comment]
> This is a useful function to have, but I'm not very keen on the name, which suggests (to me) that the function is looking for a value of the form `Some v` rather than a value for which `f` returns `Some v`.

What about `List.mapfind` or `List.map_find`?

**Comment by dbuenzli** [comment]
I agree with @yallop I think `find_map` (and a correponding `filter_map`) would be nicer. 

EDIT: base uses [`find_map`](https://b0-system.github.io/odig/doc/base/Base/List/index.html#val-find_map)

**Comment by nojb** [comment]
> I agree with @yallop I think `find_map` (and a correponding `filter_map`) would be nicer.

Since we already have `List.filter_map`, I think this is what makes the most sense.

**Comment by alainfrisch** [comment]
LGTM, modulo the naming.  We already have `filter_map`, so `find_map` would look coherent, or perhaps `find_map_opt`.

**Comment by yallop** [comment]
Agreed: I like `find_map` (or perhaps `find_map_opt`), for the analogy with the existing `filter_map`

**Comment by nojb** [comment]
> Agreed: I like `find_map` (or perhaps `find_map_opt`), for the analogy with the existing `filter_map`

I am against using `_opt` for functions returning options; rather I would use `_exn` for those raising exceptions; we did it the other way around for existing functions in the stdlib because we did not have an alternative. But for new functions I would rather not use `_opt`, even if it results in an unfortunate incoherence with existing functions.

Just my 2 cents.

**Comment by bluddy** [comment]
> But for new functions I would rather not use _opt, even if it results in an unfortunate incoherence with existing functions.

Much as I would like to achieve this goal, I think it would make the stdlib near unusable. You have to be able to immediately think of the name of the function you want. If you can't even remember if it uses the `_opt`  convention or `_exn` convention, you're screwed.

I'd prefer to have an Stdlib.V2 that fixes these things, deprecating the old stdlib.

**Comment by nojb** [comment]
> Much as I would like to achieve this goal, I think it would make the stdlib near unusable.

Maybe. In any case, I will not complain if `find_map_opt` is chosen; it was just that I find the `_opt` suffix ugly and I think it is a pity to have ugly names for commonly used functions.

**Comment by gasche** [comment]
I'm happy with `find_map`. (Initially I thought of `filter_find`, in analogy with `filter_map`, but your analogy is better than my analogy.) Thanks for the comments on the .mli, I will also integrate them.

(I don't think that `_opt` is necessary here, again in analogy with `filter_map`.)

**Comment by dbuenzli** [comment]
@bluddy Sigh. We are constantly having the same discussions about the way we should move on forward with some of the historical decision of the stdlib. 

I think it is much better if we *gradually* lead towards what we would like to have rather than continue to enshrine weird past decisions. The `_opt` suffix *reads* very weirdly if not plain wrong (e.g. `In32.of_string_opt` should really have been `Int32.option_of_string`). 

In these days of wonderful IDE completion I don't think your usability point is one and it's certainly not one as far as code reading is which as usual what should be favored.

The stdlib is already largely inconsistent and will likely ever be. People believing we'll be saved by some kind of `Stdlib.V2` design are misguided in my opinion and rather preventing us from making progress towards we would all like to have eventually.

**Comment by gasche** [comment]
Here is the `.mli` wording that I used:

> [find_map f l] applies [f] to the elements of [l] in order,
> and returns the first result of the form [Some v], or [None]
> if none exist.

(Personally I think it's obvious that the function will stop at the first suitable element, so I'm not too worried about people reading the documentation the wrong way, but I hope that this reformulation will work for @yallop and @alainfrisch.)

**Comment by gasche** [comment]
I think that stdlib additions (at least, mine) should be consensus-based rather than passing a single approval. If I understand correctly, the current status of the PR has support from @nojb, @yallop, @dbuenzli and @alainfrisch (please complain if you were misrepresented).

No specific concern with the new version has been made explicit, although there is discussion on `_opt` (I don't think we can have a broad consensus on the issue *in general*; but I feel that we may reach a consensus on this specific function -- we may already have a consensus), and I guess that the discussion on the `.mli` wording is still ongoing.

File: Changes
Comment (dbuenzli) [inline Changes]: This needs to be changed to `find_map` (as well as the title of the PR please).

File: stdlib/list.mli
Comment (dbuenzli) [inline stdlib/list.mli:238]: In general I prefer when doc strings describe the result ("`[find_map f l] is Some _` ...) than the operational behaviour ("`returns`" etc.). In that case though my trials were too convoluted and I'm personally fine with this docstring.

File: stdlib/list.ml
Comment (dbuenzli) [inline stdlib/list.ml:234]: Style: `filter_map` below has no `begin` `end` wrapping in the same case (do as you wish just saying...)

**Review by dbuenzli — APPROVED** [review]

File: Changes
Comment (gasche) [reply Changes]: Fixed, thanks.

File: stdlib/list.mli
Comment (gasche) [reply stdlib/list.mli:238]: I agree, and I agree.

**Review by dbuenzli — APPROVED** [review]
It seems little is needed to get this merged (see @nojb's comment). I again won't mention `ListLabels` because it should really die.  

A nice touch to the PR would be to add a commit that gets rid of [`Misc.Stdlib.List.find_map`](https://github.com/ocaml/ocaml/blob/c200fa14bde057623367c7605357d54029da3c2d/utils/misc.ml#L115-L121) throughout the compiler codebase.

**Comment by gasche** [comment]
I rebased the PR and made sure @nojb's comments were addressed.

(I won't merge myself; I think that stdlib additions would deserve extra carefulness in getting consensus and approval, and I don't want to speed up the process artificially.)

**Review by alainfrisch — APPROVED** [review]

**Comment by alainfrisch** [comment]
LGTM as well, will merge soon if nobody objects.  (While reviewing the interface, it occurred to me that it would make sense to move `filter_map` after `filter` in `list.mli`, but this is not really related to this PR.)

**Pull request merged**
